pkg/sync/sync_context.go

```diff
@@ -753,60 +759,71 @@ func (sc *syncContext) runTasks(tasks syncTasks, dryRun bool) runState {
 						// it is possible to get a race condition here, such that the resource does not exist when
 						// delete is requested, we treat this as a nop
 						if !apierr.IsNotFound(err) {
+							runStateMutex.Lock()
 							runState = failed
+							runStateMutex.Unlock()
 							sc.setResourceResult(t, "", common.OperationError, fmt.Sprintf("failed to delete resource: %v", err))
 						}
 					} else {
 						// if there is anything that needs deleting, we are at best now in pending and
 						// want to return and wait for sync to be invoked again
+						runStateMutex.Lock()
 						runState = pending
```

Comment:
I think apart from the data race that you are fixing here, there is also a logic error. One goroutine may set `runState` to `failed` but then another one to `pending` or the other way around. This non-deterministic behavior is probably not right, or is it? Please see https://github.com/argoproj/gitops-engine/commit/1e0e79249b8b390d0736f753a7d2d490c0a90c41 for a different approach.

Reply:
I cherry-picked from your commit.
Thank you!

https://github.com/argoproj/gitops-engine/pull/51/commits/6a762b3b226aa4aa8be9d643dd7538633d4581e5
https://github.com/argoproj/gitops-engine/pull/51/commits/c5d445f642a1283a09aa0c0056274e1e6fe17997